docs/src/lib/indexing.md

```diff
@@ -21,77 +21,146 @@ The rules for a valid type of index into a column are the following:
 * `dfr[cols]` -> a `DataFrameRow` with parent `parent(dfr)` if `cols` is a colon and `parent(dfr)[cols]` otherwise;
+
+`view` on `DataFrameRow`:
 * `@view dfr[col]` -> a `0`-dimensional view into `parent(dfr)[DataFrames.row(dfr), col]`;
 * `@view dfr[cols]` -> a `DataFrameRow` with parent `parent(dfr)` if `cols` is a colon and `parent(dfr)[cols]` otherwise;
 
+Note that when views are created with columns selector set to `:` then the view produced dynamically changes its columns'
+names and count if columns of parents are added/removed/renamed;
+
 ## `setindex!`
 
-Under construction
+The following list specifies return types of `setindex!` operations depending on argument types.
+
+In particular a description explicitly mentions if the assignment is *in-place*,
+*replaces old vectors without copying source* or *replaces old vectors with copying source*.
+
+`setindex!` on `DataFrame`:
+* `df[CartesianIndex(row, col)] = v` -> the same as `df[row, col] = v`;
+* `df[row, col] = v` -> set value of `col` in row `row` to `v` in-place;
+* `df[row, cols] = v` -> the same as `dfr = df[row, cols]; dfr[:] = v` in-place;
+* `df[rows, col] = v` -> set rows `rows` of column `col` in-place; `v` can be an abstract vector;
+* `df[rows, cols]` -> set rows `rows` of columns `cols` in-place; `v` can be an `AbstractMatrix` or `v` can be `AbstractDataFrame`
+                      (in this case column names must match);
+* `df[:, col] = v` -> the same as `df[1:nrow(df), col] = v` but replaces old vectors with a copy of source;
+                      the same as `df.col = v` if `col` is a valid identifier;
+                      also if `col` is a `Symbol` that is not present in `df` then a new column in `df` is created;
+* `df[:, cols] = v` -> the same as `df[1:nrow(df), cols] = v` but replaces old vectors with a copy of source;
+* `df[!, col] = v` -> the same as `df[1:nrow(df), col] = v` but replaces old vectors without copying source;
+                      also if `col` is a `Symbol` that is not present in `df` then a new column in `df` is created without copying of source;
+* `df[!, cols] = v` -> the same as `df[1:nrow(df), cols] = v` but replaces old vectors without copying source if `v` is an `AbstractDataFrame`.
+
+`setindex!` on `SubDataFrame`:
+* `sdf[CartesianIndex(row, col)] = v` -> the same as `sdf[row, col] = v`;
+* `sdf[row, col]` -> set value of `col` in row `row` to `v` in-place;
+* `sdf[row, cols]` -> the same as `dfr = df[row, cols]; dfr[:] = v` in-place;
+* `sdf[rows, col]` -> set rows `rows` of column `col` in-place; `v` can be an abstract vector;
+* `sdf[rows, cols]` -> set rows `rows` of columns `cols` in-place;
+                       `v` can be an `AbstractMatrix` or `v` can be `AbstractDataFrame` when column names must match;
+* `sdf[:, col]` -> the same as `sdf[1:nrow(df), col] = v`; the same as `sdf.col = v` if `col` is a valid identifier;
```

Comment:
These are somewhat inconsistent with what happens with `DataFrame`, where columns are replaced. Maybe we should require using `1:end` or `.=`?

Reply:
I had a problem what to do with these cases. The rules are inconsistent, but it is not possible to make them consistent (we can do only in-place operation).
I have added these rules because you can have code and then add `@view` in front of it later. Without these rules such an operation would break the code.

In order to be consistent we would have to drop special handling of `:` and `!` for `DataFrame`, but I understand we do not want this.

Reply:
How about making these throw an error that points to `1:end` and `.=` for now? It sounds like a good idea to force users to use a different syntax from `DataFrame` here since the result is completely different. They should use a syntax that does in-place column setting both on `DataFrame` and `SubDataFrame`.

Reply:
OK - we can disallow it. We can always introduce it later if needed (I will update the docs for this case). In general setting whole columns of `SubDataFrame` is I guess a corner case that rarely happens in practice.

The only problem is that it covers `sdf.col = v` case which we should allow and it will be inconsistent (`sdf.col = v` is in place while `df.col = v` replaces with copy).

Also we still have an inconsistency that `df[:, col] .= v` allocates fresh columns while `sdf[:, col] .= v` is in place.

The problem is that in order to make this consistent we would have to make `:` in-place always. This is doable, but then we should have this `setcol!` function for column replacement and drop the whole `!` idea as it will be redundant then.

So the choice is:
1. use current design with `:` and `!` as it does what users expect most of the time (but either gives some inconsistency or is disallowed in some cases)
2. be strict with consistency and add `setcol!`

I think option 1. is better (we will still have inconsistency between `df.col = v` and `sdf.col = v` anyway if we go for option 2.). Then the question is if in option 1. we minimize inconsistency (by disallowing certain operations - your current proposal) or simply accept the inconsistency (this is the current state in the docs).